src/dbSta/src/dbNetwork.cc

```diff
@@ -121,36 +126,66 @@ class DbInstanceChildIterator : public InstanceChildIterator
 
  private:
   const dbNetwork* network_;
+  dbModule* module_;
   bool top_;
-  dbSet<dbInst>::iterator iter_;
-  dbSet<dbInst>::iterator end_;
+  dbSet<dbInst>::iterator dbinst_iter_;
+  dbSet<dbInst>::iterator dbinst_end_;
+  dbSet<dbModInst>::iterator modinst_iter_;
+  dbSet<dbModInst>::iterator modinst_end_;
 };
 
 DbInstanceChildIterator::DbInstanceChildIterator(const Instance* instance,
                                                  const dbNetwork* network)
     : network_(network)
 {
   dbBlock* block = network->block();
+  module_ = block->getTopModule();
+  modinst_iter_ = ((dbModule*) module_)->getModInsts().begin();
+  modinst_end_ = modinst_iter_;
+  dbinst_iter_ = ((dbModule*) module_)->getInsts().begin();
+  dbinst_end_ = dbinst_iter_;
+
   if (instance == network->topInstance() && block) {
-    dbSet<dbInst> insts = block->getInsts();
+    module_ = block->getTopModule();
     top_ = true;
-    iter_ = insts.begin();
-    end_ = insts.end();
+    modinst_iter_ = module_->getModInsts().begin();
+    modinst_end_ = module_->getModInsts().end();
+    dbinst_iter_ = module_->getInsts().begin();
+    dbinst_end_ = module_->getInsts().end();
   } else {
     top_ = false;
+    // need to get module for instance
+    dbInst* db_inst = nullptr;
+    dbModInst* mod_inst = nullptr;
+    network->staToDb(instance, db_inst, mod_inst);
+    if (mod_inst) {
+      module_ = mod_inst->getMaster();
+      modinst_iter_ = module_->getModInsts().begin();
+      modinst_end_ = module_->getModInsts().end();
+      dbinst_iter_ = module_->getInsts().begin();
+      dbinst_end_ = module_->getInsts().end();
+    }
   }
 }
 
 bool DbInstanceChildIterator::hasNext()
 {
-  return top_ && iter_ != end_;
+  return !((dbinst_iter_ == dbinst_end_) && (modinst_iter_ == modinst_end_));
 }
 
 Instance* DbInstanceChildIterator::next()
 {
-  dbInst* child = *iter_;
-  iter_++;
-  return network_->dbToSta(child);
+  Instance* ret = nullptr;
+  if (dbinst_iter_ != dbinst_end_) {
+    dbInst* child = *dbinst_iter_;
+    dbinst_iter_++;
+    ret = network_->dbToSta(child);
+  } else if (modinst_iter_ != modinst_end_) {
+    dbModInst* child = *modinst_iter_;
+    modinst_iter_++;
```

Comment:
When you change modules do you need to reset dbinst_iter_  to that module's children?

Reply:
This iterator is traversing the children of an instance not the full hierarchy. There are two types of children: module instance and dbInstance. The InstanceChildIterator is called recursively in the STA to effect a full hierarchical traversal -- see for example Network.cc/findNetsHierMatching. So we don't need to reset the dbinst_iter_ to push down into the module instances children.